### ISSUE TITLE
`denote-rename-file` and `denote-rename-file-using-front-matter` behave incorrectly in `org-capture` buffer

### ISSUE THREAD

**Issue by jeffvalk** [protesilaos/denote#500]
First of all, thanks for Denote and all the thought that has gone into it!

Here is what I'm seeing...

### Unexpected behavior

When in an `org-capture` buffer:

- Invoking `denote-rename-file` produces an unexpected directory prompt, and
- Invoking `denote-rename-file-using-front-matter` errors with `(wrong-type-argument stringp nil)`.

I think these functions do not check whether the buffer currently has a file (e.g. `denote--file-type-org-extra-p`).

### Steps to reproduce with `emacs -Q`:

```elisp
(require 'denote "path/to/denote-3.1.0/denote.el")

;; example org-capture template from manual
(with-eval-after-load 'org-capture
  (add-to-list 'org-capture-templates
               '("n" "New note (with Denote)" plain
                 (file denote-last-path)
                 #'denote-org-capture
                 :no-save t
                 :immediate-finish nil
                 :kill-buffer t
                 :jump-to-captured t)))

```

1. `M-x org-capture`, and select "New note (with Denote)"
2. Prior to finishing the capture with `org-capture-finalize` (`C-c C-c`), call either of Denote's rename commands:

   - `M-x denote-rename-file` brings up a minibuffer prompt: "Rename FILE Denote-style: "
   - `M-x denote-rename-file-using-front-matter` produces an error: `(wrong-type-argument stringp nil)`

### Expected result

I would expect the following behavior in an `org-capture` buffer, which is not yet saved to a file:

- `denote-rename-file`
   - Prompt according to `denote-prompts`
   - Update front matter
   - Update the value of `denote-last-path`

- `denote-rename-file-using-front-matter` 
   - Update `denote-last-path`

Then when `org-capture-finalize` is called, the buffer should be written by `org-capture` to a file with the updated file name stored in `denote-last-path`.

**Comment by jeanphilippegg** [comment]
Thank you for the detailed issue!

> Then when `org-capture-finalize` is called, the buffer should be written by `org-capture` to a file with the updated file name stored in `denote-last-path`.

I have tried to make renaming commands work in org-capture buffer on my fork (branch `denote-org-capture`), but it seems like updating `denote-last-path` in `denote-rename-file` is too late, meaning that the initial path is always used.

If you or Prot knows how to update the destination path after the end of `denote-org-capture`, then this can be supported. Otherwise, it will probably stay a limitation. We can still make the code output a better error message though.

**Comment by protesilaos** [comment]
> From: Jean-Philippe Gagné Guay ***@***.***>
> Date: Sat, 21 Dec 2024 21:04:24 -0800
>
> Thank you for the detailed issue!
>
>> Then when `org-capture-finalize` is called, the buffer should be
>> written by `org-capture` to a file with the updated file name stored
>> in `denote-last-path`.
>
> I have tried to make renaming commands work in org-capture buffer on
> my fork (branch `denote-org-capture`), but it seems like updating
> `denote-last-path` in `denote-rename-file` is too late, meaning that
> the initial path is always used.
>
> If you or Prot knows how to update the destination path after the end
> of `denote-org-capture`, then this can be supported. Otherwise, it
> will probably stay a limitation. We can still make the code output a
> better error message though.

I am not aware of a way to circumvent this. Even if we do not use the
'denote-last-path' in the template, we still need a function which will
ultimately do about the same. Per the doc string of 'org-capture-templates':

    (function function-finding-location)
       Most general way: write your own function which both visits
       the file and moves point to the right location

But even then, the Org capture buffer is not associated with its target
file. Unless there is some internal state to how org-capture knows the
target file, I think there is nothing for us to do. (And even if
org-capture keeps track of the file internally, it will probably be a
bad idea for us to mess around with it via some advice or whatnot.)

-- 
Protesilaos Stavrou
https://protesilaos.com

**Comment by jeanphilippegg** [comment]
I had edited my previous comment with a possible solution.

I made pull request #502. Changing `denote-last-path` is ignored after `denote-org-capture`, but we can keep track of the modifications and use `org-capture-after-finalize-hook` to rename the destination file at the end of a capture.

It works on my end. If you think it complicates the code too much or that there may be other issues, feel free to not merge it!
